include/boost/math/special_functions/jacobi_theta.hpp

```diff
@@ -0,0 +1,616 @@
+template <class RealType, class Policy>
+inline RealType
+jacobi_theta2tau(RealType z, RealType tau, const Policy& pol)
+{
+    BOOST_MATH_STD_USING
+    unsigned n = 0;
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    RealType q_n, delta, result = RealType(0);
+
+    if (tau <= 0.0) {
+        return policies::raise_domain_error<RealType>("boost::math::jacobi_theta2tau<%1%>(%1%)",
+                "tau must be greater than 0 but got %1%.", tau, pol);
+    } else if (tau < 1.0 && abs(z) == 0.0) {
+        return jacobi_theta4tau(z, 1/tau, pol) / sqrt(tau);
+    } else if (tau < 1.0) { // DLMF 20.7.31
+        z = fmod(z, constants::two_pi<RealType>());
+        while (z > constants::pi<RealType>()) {
+            z -= constants::two_pi<RealType>();
+        }
+        while (z < -constants::pi<RealType>()) {
+            z += constants::two_pi<RealType>();
+        }
+
+        return _IMAGINARY_jacobi_theta4tau(z, 1/tau, pol);
+    }
+
+    do {
+        q_n = exp(-tau * constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5));
+        delta = q_n * cos(RealType(2*n+1)*z);
+        result += delta + delta;
+        n++;
+    } while (!_jacobi_theta_converged(result, q_n, eps));
+
+    return result;
+}
+
+template <class T>
+inline T jacobi_theta2tau(T z, T tau) {
+    return jacobi_theta2tau(z, tau, policies::policy<>());
+}
+
+// Second Jacobi theta function, parameterized by q
+// = 2 * Σ q^(n+1/2)^2 * cos((2n+1)z)
+template <class RealType, class Policy>
+inline RealType
+jacobi_theta2(RealType z, RealType q, const Policy& pol) {
+    BOOST_MATH_STD_USING
+    if (q <= 0.0 || q >= 1.0) {
+        return policies::raise_domain_error<RealType>("boost::math::jacobi_theta2<%1%>(%1%)",
+                "q must be greater than 0 and less than 1 but got %1%.", q, pol);
+    }
+    return jacobi_theta2tau(z, -log(q)/constants::pi<RealType>(), pol);
+}
+
+template <class T>
+inline T jacobi_theta2(T z, T q) {
+    return jacobi_theta2(z, q, policies::policy<>());
+}
+
+// Third Jacobi theta function, minus one (Parameterized by tau - assumed imaginary)
+// This function preserves accuracy for small values of q (i.e. |q| < exp(-π) ≅ 0.043)
+// For larger values of q, the minus one version usually won't help.
+// = 2 * Σ exp(iπτ*(n)^2) * cos(2nz)
+template <class RealType, class Policy>
+inline RealType
+jacobi_theta3m1tau(RealType z, RealType tau, const Policy& pol)
+{
+    BOOST_MATH_STD_USING
+
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    RealType q_n, delta, result = RealType(0);
+    unsigned n = 1;
+
+    if (tau < 1.0)
+        return jacobi_theta3tau(z, tau, pol) - RealType(1);
+
+    do {
+        q_n = exp(-tau * constants::pi<RealType>() * RealType(n)*RealType(n));
+        delta = q_n * cos(RealType(2*n)*z);
+        result += delta + delta;
+        n++;
+    } while (!_jacobi_theta_converged(result, q_n, eps));
+
+    if (abs(result) < eps * eps)
+        return RealType(0);
```

Comment:
I don't understand why this needs to be done. There are plenty of floating point numbers less than eps*eps; they are perfectly valid outputs aren't they?

Reply:
You are correct; I removed similar checks after previous discussion, but forgot to take this one out as well.